- Description: qeth: potential race in debug feature
  Symptom:     With multiple CPUs it is possible that inconsistent debug
               output in the s390dbf is produced.
  Problem:     We had only one global char buffer used by the macro
               QETH_DBF_TEXT_. This buffer can potentially be used concurrently
               by multiple CPUs.
  Solution:    Defined the char buffer qeth_dbf_txt_buf per CPU. Modified the
               debug macro to use get_cpu_var and put_cpu_var.
  Problem-ID:  --
  Archs:       s390-31, s390-64

--- linux-2.5/drivers/s390/net/qeth.h	24 May 2004 15:46:48 -0000	1.98.2.9
+++ linux-2.5/drivers/s390/net/qeth.h	2 Jun 2004 07:35:20 -0000	1.98.2.10
@@ -23,7 +23,7 @@
 
 #include "qeth_mpc.h"
 
-#define VERSION_QETH_H 		"$Revision: 1.98.2.9 $"
+#define VERSION_QETH_H 		"$Revision: 1.98.2.10 $"
 
 #ifdef CONFIG_QETH_IPV6
 #define QETH_VERSION_IPV6 	":IPv6"
@@ -91,10 +91,14 @@
 		debug_event(qeth_dbf_##name,level,(void*)(addr),len); \
 	} while (0)
 
-#define QETH_DBF_TEXT_(name,level,text...)				  \
-	do {								  \
-		sprintf(qeth_dbf_text_buf, text);			  \
-		debug_text_event(qeth_dbf_##name,level,qeth_dbf_text_buf);\
+extern DEFINE_PER_CPU(char[256], qeth_dbf_txt_buf);
+
+#define QETH_DBF_TEXT_(name,level,text...)				\
+	do {								\
+		char* dbf_txt_buf = get_cpu_var(qeth_dbf_txt_buf);	\
+		sprintf(dbf_txt_buf, text);			  	\
+		debug_text_event(qeth_dbf_##name,level,dbf_txt_buf);	\
+		put_cpu_var(qeth_dbf_txt_buf);				\
 	} while (0)
 
 #define QETH_DBF_SPRINTF(name,level,text...) \

--- linux-2.5/drivers/s390/net/qeth_main.c	24 May 2004 16:06:47 -0000	1.77.2.16
+++ linux-2.5/drivers/s390/net/qeth_main.c	2 Jun 2004 07:35:20 -0000	1.77.2.17
@@ -1,6 +1,6 @@
 /*
  * 
- * linux/drivers/s390/net/qeth_main.c ($Revision: 1.77.2.16 $)
+ * linux/drivers/s390/net/qeth_main.c ($Revision: 1.77.2.17 $)
  *
  * Linux on zSeries OSA Express and HiperSockets support
  *
@@ -12,7 +12,7 @@
  *			  Frank Pavlic (pavlic@de.ibm.com) and
  *		 	  Thomas Spatzier <tspat@de.ibm.com>
  *
- *    $Revision: 1.77.2.16 $	 $Date: 2004/05/24 16:06:47 $
+ *    $Revision: 1.77.2.17 $	 $Date: 2004/06/02 07:35:20 $
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
@@ -78,7 +78,7 @@
 #include "qeth_mpc.h"
 #include "qeth_fs.h"
 
-#define VERSION_QETH_C "$Revision: 1.77.2.16 $"
+#define VERSION_QETH_C "$Revision: 1.77.2.17 $"
 static const char *version = "qeth S/390 OSA-Express driver";
 
 /**
@@ -91,7 +91,8 @@
 static debug_info_t *qeth_dbf_trace = NULL;
 static debug_info_t *qeth_dbf_sense = NULL;
 static debug_info_t *qeth_dbf_qerr = NULL;
-static char qeth_dbf_text_buf[255];
+
+DEFINE_PER_CPU(char[256], qeth_dbf_txt_buf);
 
 /**
  * some more definitions and declarations

